Enhance Provider to Support bare OS installations and Use Terraform Plugin Framework

TheNilesh opened the pull request:
- Migrate the provider from the SDKv2 to the Terraform Plugin Framework to take advantage of the following benefits:
  - Improved development experience
  - Better testability
  - Simplified configuration schema

- Tailor the provider for the bareOS installation, removing support for AWS and other cloud providers.

- Support cluster attributes introduced later

- Implement the following data sources to improve programmability:
  - `node` data source and filters
  - `host` data source
  - `cluster` data source and filters

- Implement functionality to generate a kubeconfig file from Terraform, enabling better integration

TheNilesh committed:
Set baseline for migration to plugin-framework

In this commit:
- Initialize the project as per guideline of plugin-framework.
- Generate empty code files as placeholders.
- Establish a baseline for further development.

This is the starting point for migrating from SDKv2 to the plugin framework.
Further development and implementation will build upon this baseline.

TheNilesh committed:
Integrate qbert client

- Integrate qbert client from pf9-sdk
- Generate documentation
- Add examples
- Support most of the attributes

TheNilesh committed:
Update documentation and development setup instructions

TheNilesh committed:
Fix attributes

TheNilesh committed:
Add attributes to cluster resource

TheNilesh committed:
use state for unknown

TheNilesh committed:
Add etcd attributes and attach node issue fix

TheNilesh committed:
Validate kube role version

TheNilesh committed:
Fix kuberoleversion support
- Upgrade pf9-sdk-go
- Support 70 new attributes
- Fix tags, interface-detection attributes

TheNilesh committed:
Documentation with examples

TheNilesh committed:
Support addons

TheNilesh committed:
Code optimizations

TheNilesh committed:
Fix issues with addon marshalling

TheNilesh committed:
Implement clusters and kubeconfig data source

- Improve addon patch code

TheNilesh committed:
Add pf9_kubeconfig and pf9_clusters data sources

TheNilesh committed:
Support tenant attribute in cluster filters

TheNilesh committed:
Addons schema changes

TheNilesh committed:
Plan modifier for kube role version

TheNilesh committed:
Failed attempt to support addons

TheNilesh committed:
Improve addons schema

* Improve kubeconfig data source schema
* Add validations, fixes to kube role version and upgrade

TheNilesh committed:
Add host data source and fix upgrade cluster errors

TheNilesh committed:
Update data source documentation and add new examples

TheNilesh committed:
Improve schema using nested attributes
- Add requires replace for every non-updateable attribute

TheNilesh committed:
Update etcd backup schema in docs

TheNilesh committed:
Add etcd, calico_ipv4 attribute to pf9_cluster resource

TheNilesh committed:
Fix cluster upgrade issue and add node verification in cluster creation

TheNilesh committed:
Provider configuration and catapult monitoring flag

- clusters data source can now work without filters(to list all  clusters)
- provider only accepts configuration through attributes not through env variables

mithilarun reviewed (commented):
@TheNilesh I don't really understand the newer architecture, so I'm not the best person to be reviewing this. Regardless, can you update the PR description with notes on the tests you ran to validate this?

TheNilesh commented:
Hi @mithilarun I have raised a new pull request due to one technical issue. Please review: https://github.com/platform9/terraform-provider-pf9/pull/29

I will be addressing your comment in the new PR.
